pageserver/src/walredo.rs

```diff
@@ -333,7 +334,9 @@ impl PostgresRedoManager {
                 // and hence the current `apply_wal_records()` calls will observe
                 //  `output.stdout.as_raw_fd() != stdout_fd` .
                 if let Some(proc) = self.stdin.lock().unwrap().take() {
-                    proc.child.kill_and_wait();
+                    if stdin_fd == proc.stdin.as_raw_fd() {
+                        proc.child.kill_and_wait();
+                    }
```

Comment:
Comparing raw FD doesn't seem bullet-proof. A file descriptor is just an integer, and the kernel reuses them. If a file/pipe is closed and opened again, you might get the same raw file descriptor again. A straightforward fix would be to add explicit restart counter to ProcessInput, and compare that instead.

Is it OK that we have 'taken' the self.stdin field, and we just leak it here without killing the process?

Reply:
There is another comparison of raw FDs here: https://github.com/neondatabase/neon/blob/21deb81acbb0516e2f8d3934f40f951b8c7d3345/pageserver/src/walredo.rs#L894. I thnk that can also get fooled if the new FD happens to be the same as old one.

Reply:
I suggest to introduce another struct to tie together the stdin/stdout/stderr of a particular launch of the process, along the lines of:

```
--- a/pageserver/src/walredo.rs
+++ b/pageserver/src/walredo.rs
@@ -120,8 +120,12 @@ pub struct PostgresRedoManager {
     #[cfg(feature = "testing")]
     dump_sequence: AtomicUsize,
 
-    stdout: Mutex<Option<ProcessOutput>>,
-    stdin: Mutex<Option<ProcessInput>>,
-    stderr: Mutex<Option<ChildStderr>>,
+    proc: Option<Arc<WalRedoProcess>>,
+}
+
+struct WalRedoProcess {
+    stdout: Mutex<ProcessOutput>,
+    stdin: Mutex<ProcessInput>,
+    stderr: Mutex<ChildStderr>,
 }
 
```

I think you'll need another lock on 'proc' to protect the reference, or maybe use `atomic_ref`. But that's the general idea.

Reply:
I had a technical argument with Konstantin about relying on the file descriptor lifecycles  many months ago (read [starting here](https://github.com/neondatabase/neon/pull/3368#pullrequestreview-1263240030)).
I budged at the time, I intend to budge a second time. Let's keep this PR minimal so we make progress in root-causing the issue that we're investigating. We can have a separate discussion about better walredo.rs design.

Reply:
Ok. I'd still suggest that we at least use a add 'restart counter' field instead of relying on the fds. That's a very narrow change. Looking at the code as it is, I don't trust that the fd might not get reused, and an explicit field seems more readable too.

Reply:
Opened https://github.com/neondatabase/neon/pull/5549 for that, on top of this PR. It can be done separately, but I wouldn't mind sneaking it in here either.

Reply:
We have discussed this "hack" with checking field descriptors long time ago.
And there is large comment why it safe to do it:)
May be it is better to add `restart_counter` to readability, but actually I do not understand why comparing descriptors is not safe (if there is a warranty that them aren't closed).